### ISSUE TITLE
BranchPtr and users are unsound

### ISSUE THREAD

**Issue by ColonelThirtyTwo** [y-crdt/y-crdt#506]
`BranchPtr` and its wrappers (The `TypeRef` objects) are unsound. They provide unchecked simultaneous `&mut` access to a single `Branch` object at a time, which, itself, is allocated as `Arc<Branch>` providing only immutable access. Even trivial usage of the crate causes errors in miri:

```rust
use yrs::{branch::Branch, Doc, Text as _, Transact};

fn main() {
    let doc = Doc::new();
    let item = doc.get_or_insert_text("test");
    // Miri fails in the above line. The remaining code shows some examples of UB access

    let branch_ptr = AsRef::<Branch>::as_ref(&item).id().get_branch(
        &doc.transact()
    ).unwrap();
    std::mem::drop(doc); // item and branch_ptr are now now dangling

    branch_ptr.id(); // UB: branch_ptr is dangling

    let doc = Doc::new();
    let txn = doc.transact();
    item.len(&txn); // UB: access to dangling item
}

```

```
> cargo +nightly miri run
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.02s
     Running `/home/col/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/cargo-miri runner target/miri/x86_64-unknown-linux-gnu/debug/test-pycrdt-sound`
error: Undefined Behavior: trying to retag from <31969> for Unique permission at alloc12689[0x10], but that tag only grants SharedReadOnly permission for this location
   --> /home/col/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/non_null.rs:403:18
    |
403 |         unsafe { &mut *self.as_ptr() }
    |                  ^^^^^^^^^^^^^^^^^^^
    |                  |
    |                  trying to retag from <31969> for Unique permission at alloc12689[0x10], but that tag only grants SharedReadOnly permission for this location
    |                  this error occurs as part of retag at alloc12689[0x10..0x90]
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <31969> was created by a SharedReadOnly retag at offsets [0x10..0x88]
   --> src/main.rs:5:16
    |
5   |     let item = doc.get_or_insert_text("test");
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: BACKTRACE (of the first span):
    = note: inside `std::ptr::NonNull::<yrs::branch::Branch>::as_mut::<'_>` at /home/col/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/non_null.rs:403:18: 403:37
    = note: inside `<yrs::branch::BranchPtr as std::ops::DerefMut>::deref_mut` at /home/col/.cargo/registry/src/index.crates.io-6f17d22bba15001f/yrs-0.21.2/src/branch.rs:97:18: 97:33
    = note: inside `yrs::Store::get_or_create_type::<std::sync::Arc<str>>` at /home/col/.cargo/registry/src/index.crates.io-6f17d22bba15001f/yrs-0.21.2/src/store.rs:135:17: 135:32
    = note: inside `yrs::Root::<yrs::TextRef>::get_or_create::<yrs::TransactionMut<'_>>` at /home/col/.cargo/registry/src/index.crates.io-6f17d22bba15001f/yrs-0.21.2/src/branch.rs:706:22: 706:80
    = note: inside `yrs::Doc::get_or_insert_text::<&str>` at /home/col/.cargo/registry/src/index.crates.io-6f17d22bba15001f/yrs-0.21.2/src/doc.rs:191:9: 191:68
note: inside `main`
   --> src/main.rs:5:16
    |
5   |     let item = doc.get_or_insert_text("test");
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error
```

These issues can cause bugs, as the Rust optimizer assumes that unsafe code properly accesses pointers and obeys Rust's aliasing rules, which yrs does not.

There's probably other similar issues with the yrs codebase as well - this one is just the most obvious one I noticed due to `BranchPtr`'s name and the fact that it's a wrapper for a `NonNull<Branch>` and nothing else to track or restrict its usage.

**Comment by Horusiath** [comment]
Yes, it's an old artifact from times when yrs primary reason to be was to serve as cross-language binding layer. It sticked to the core implementation since then and nowadays it's hard to change without a lot of (currently unpaid) effort and breaking a significant parts of the API for existing users.

In practice this is rarely an issue - most of the shared types operations requires transaction access which can only be done if you have a live `Doc` around. In case when you need to provide a safe access there's a [Hook](https://docs.rs/yrs/latest/yrs/branch/struct.Hook.html) struct that is basically a purely logical descriptor of a specific branch, can be serialized and materialised into specific shared ref instance on demand.

**Comment by ColonelThirtyTwo** [comment]
> In practice this is rarely an issue - most of the shared types operations requires transaction access which can only be done if you have a live Doc around

Object liveness is only part of the issue. As miri found, it also violates Rust's aliasing rules, which can cause invalid optimizations even if the objects are live.

**Comment by Horusiath** [comment]
> As miri found, it also violates Rust's aliasing rules, which can cause invalid optimizations even if the objects are live.

Except it has never happened in practice so far - and there are plenty of users for this library. Conservative static analysis can produce false negatives and borrow checker is the best example of that. In future this indeed may be a problem and I'm willing to put work in case when we'll confirm that this is in fact the case.

It's just that atm. I cannot afford to put necessary time into it. I don't mind reviewing the PR of such fix.

**Comment by ColonelThirtyTwo** [comment]
> It's just that atm. I cannot afford to put necessary time into it.

I'm not trying to downplay this, but...

> Conservative static analysis can produce false negatives and borrow checker is the best example of that.

Treating the Rust borrow checker as a "static analysis tool" to be disregarded at the slightest bit of inconvenience is pretty antithetical to Rust's philosophy and undermines the safety it's trying to provide. Rust references still have aliasing rules even in unsafe code, and the expectation that Rust prescribes is that one using a crate cannot trigger UB without using an `unsafe` API.

And I have found memory access issues in yrs, which is concerning in the context of a data format where untrusted users are sending data to servers and other clients. (PS. how do you want me to disclose it, once I've done some investigation?)

> I don't mind reviewing the PR of such fix.

I may try to do some investigation into fixing issues, but what will you say to fixes that require deep changes to the API?

The most straightforward way to fix `BranchPtr` would be to turn it into a reference retrieved from a transaction, but that also means that it can't outlive the transaction, so you can't keep it around afterwards. I personally don't mind that - having a reference to a thing inside of a doc but outside of any locking of it is weird IMO - but it also isn't how the JS implementation works.

Alternatives would be have `BranchPtr` wrap an `Arc<Mutex<Branch>>` or some other locking or have `BranchPtr` store a non-pointer way of referring to a specific object, both of which have tradeoffs.

**Comment by Horusiath** [comment]
> And I have found memory access issues in yrs for loading updates, which is concerning in the context of a data format where untrusted users are sending data to servers and other clients. 

You mean ie. when trying to access deallocated `MapRef` etc. Or something else?

> Alternatives would be have BranchPtr wrap an Arc<Mutex<Branch>> or some other locking or have BranchPtr store a non-pointer way of referring to a specific object, both of which have tradeoffs.

Not going to happen. It's way too expensive. There are people who use Yrs in game engines, and even a single mutex in transaction access was causing them a headache.

The first option (resolving `&'txn Branch`) is a valid approach though. You don't need to keep it around: all binding languages nowadays (including ywasm) use BranchIDs and then resolving specific branches from current transaction context or throw an exception if that's not possible.

This however will involve quite a bit of cleaning ie. removing transaction parameter from all of the operations. If we want to be safer than the current implementation, we should cache transaction reference in the `&Branch` wrapper, otherwise user could by accident pass transaction from another doc. The problem is: this basically changes every shared ref method. In the past we had much smaller API break (between version 0.18->0.19) which took maintainers months to migrate to and from what I've personally experienced some companies haven't upgraded it to this day.

**Comment by ColonelThirtyTwo** [comment]
> You mean ie. when trying to access deallocated `MapRef` etc. Or something else?

No. As in maliciously crafted `Update`s.

**Comment by Horusiath** [comment]
Do you have some reproducible scenario? It's a first time I've heard about this.

**Comment by ColonelThirtyTwo** [comment]
Again, how do you want me to disclose it? I'd rather not post about a potentially exploitable bug on a public issue tracker.

**Comment by Horusiath** [comment]
Could you send me an e-mail? (It's available in my profile)

**Comment by djc** [comment]
@Horusiath the OP submitted an issue to the advisory-db:

- https://github.com/rustsec/advisory-db/issues/3118

In it, they claim that the project is "very unsound". The advisory-db generally prefers to let maintainers decide whether an advisory is published -- could you comment on the why this issue has not been fixed in the past years? It seems there's still quite a lot of activity and in my opinion fixing unsound API should generally be high priority.
